pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java

```diff
@@ -235,20 +238,33 @@ private CompletableFuture<Reader<T>> readAllExistingMessages(Reader<T> reader) {
         AtomicLong messagesRead = new AtomicLong();
 
         CompletableFuture<Reader<T>> future = new CompletableFuture<>();
-        readAllExistingMessages(reader, future, startTime, messagesRead);
+        reader.getLastMessageIdsAsync().thenAccept(lastMessageIds -> {
+            Optional<TopicMessageId> optionalTopicMessageId = lastMessageIds.stream().max(Comparator.naturalOrder());
```

Comment:
Some things to concern:

First case: working on partitioned topics:
 Imagine:
- `topic-partiton-1` - containing 1 million messages
- `topic-partition-2` - containing 1 message

When a producer writes a new message to topic-partition-2 and it gets to TableView, the method will be stopped.


Second case: work with compaction 
- After the compaction, if the new client connects, the messages read first will have messageIds from the compacted ledger. Will compareTo work in that case?

Third case:
- Comparison of MessageId when delivered in batches (this should probably be OK now)


I think that the algorithm should store every last message from each partition (first issue). Then the stop mechanism should be based on the exact message delivery.

Reply:
> Second case: work with compaction 
> - After the compaction, if the new client connects, the messages read first will have messageIds from the compacted ledger. Will compareTo work in that case?

Messages read from a compacted topic by a normal client will be no different to messages read from a non-compacted topic, though there will be gaps in the message IDs. See https://github.com/apache/pulsar/wiki/PIP-14:-Topic-compaction

Reply:
> I think that the algorithm should store every last message from each partition (first issue). Then the stop mechanism should be based on the exact message delivery.

I totally agree. We only ensure that messages in the same partition are ordered, so the strategy of only judging the largest message ID is wrong. Thanks for your reminder.

Reply:
Just to consider:
If the compacted ledger is created after the ledger that holds the backlog, it may have the higher ledger Id, right?

- So you reed the latest message from the backlog
- You start reading from the beginning (compacted ledger, first message)
- The compacted ledger has higher ID, therefore the MessageId.equals won't work as expected

Is there a mechanism to prevent that?